Source/DafnyLanguageServer/Workspace/ChangeProcessors/IRelocator.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections.Generic;
+using Microsoft.Dafny.LanguageServer.Language.Symbols;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Threading;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.ChangeProcessors {
+  /// <summary>
+  /// Implementations of this interface are responsible to relocate symbols and diagnostics
+  /// to their new positions according to the given <see cref="DidChangeTextDocumentParams"/>.
+  /// </summary>
+  public interface IRelocator {
```

Comment:
We may want to consider keeping these two relocation mechanisms in distinct classes since they're unrelated, at least separate interfaces.

Reply:
Actually, I don't understand why we're using interfaces that have only one implementation. I would understand if we're stubbing the interfaces in testing, but we're not. What benefit is gained from having `IRelocator` ?

Reply:
I agree that it's over-engineered to have an interface for everything while there is only one implementation.
It's a usual convention for (ASP).NET Core applications to create an interface and only inject these rather than the concrete type. This helps testing since the interfaces can be mocked. However, due to the strong dependence on `Program`, things got difficult to unit test separately hence why there mostly only integration tests.

One thing I have in mind is to create an interface for `DafnyDocument` and its dependencies to make unit testing easier in the future. I didn't do that from the beginning because one of my goals when I initially created the language server was to not touch the compiler code base.